Add Support for Fine-Tuning Foundation Models with `--from-foundation` and add `CheMeleon`

JacksonBurns opened the pull request:
## Description

This pull request implements support for finetuning foundation models in `ChemProp` via the Command Line Interface (CLI), which was not previously possible.

A demo notebook for the same, but via the Python module interface, is also added.

Finally, support for the [`CheMeleon`](https://github.com/JacksonBurns/chemeleon) model by myself and @akshatzalte is also added.

## Example / Current workflow

Right now, ChemProp supports transfer learning in which a model is trained and then fine-tuned on the _same_ task. This does **not** allow the user to change the architecture during finetuning.

It is desirable to re-use only the learned representation and to create a new FFN that applies to some entirely new task. This is currently not possible via the CLI for ChemProp. It is possible via Python Modules, but we don't have a demo for it.

## Bugfix / Desired workflow

This PR exposes finetuning using a pretrained learned representation through a new CLI flag `--from-foundation`. Passing the name of a foundation model with this flag will automatically download it from Zenodo and then initialize the Message Passing, featurizer, and Aggregation according to that model. Documentation for this new feature has also been added.

The demo notebook shows how to do this as well, basically just a formalization of what was already possible via the Python module interface. This demo notebook has also been added to the documentation.

This PR also serves as a demo for how to add other foundation models in the future. Any foundation models added in the future should follow the below process (click the link to see the representative commit):
 1. 52b5f3b0c1a2b6515482de09daa8f1a226f1d731 add the model to the enumeration, then add specific logic to validate arguments, then implement the downloading and initialization of the pretrained model following the example of `CheMeleon`
 2. 245cd6487e2a581a3e3354829e34f37b82dfd94d add (or update an existing) demo notebook for the new model
 3. 24e906f5803ccc186fefb689a8ad6d7938aaebbb add the model the list of available models in the CLI

Any newly added foundation model will be automatically tested by `pytest`.

## Questions

We can change the name of the flag, if desired.

## Relevant issues

This was an MLPDS feature request from @am2145 as well, though I never opened an issue to track it (oops).

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

akshatzalte reviewed (commented):
LGTM

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -145,6 +153,10 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         help="Freeze the first ``n`` layers of the FFN from the checkpoint model (specified by ``--checkpoint``). The message passing layer should also be frozen with ``--freeze-encoder``.",
     )
+    transfer_args.add_argument(
+        "--from-foundation",
```
at the momentit doesn't support loading from local foundation models easily... in order to use this flag you have to actually modify the source to add the logic for loading it

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -497,6 +509,43 @@ def validate_train_args(args):
         )
 
+    if (fm_name := args.from_foundation) is not None:
+        try:
```
Doesn't work because `args.from_foundation` is of type `str` - in order to check if it is `in FoundationModels` we need to first look up its enum type with `FoundationModels.get`

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -497,6 +509,43 @@ def validate_train_args(args):
+        except KeyError:
+            raise ArgumentError(
+                argument=None,
+                message=f"Unrecognized foundation model name {fm_name}! Should be one of: {', '.join((FoundationModels.keys()))}",
```
I've actually intentionally not sorted here - this way the enum will be printed in the same order that the models are named in the documentation

JacksonBurns commented:
@akshatzalte thanks for the quick review. I've replied to your comments and changed the maximum complexity line in the flake8 file - let me know how it looks.

I've also scrolled through the new docs and everything looks correctly rendered.

akshatzalte commented:
LGTM

KnathanM commented on examples/chemeleon_foundation_finetuning.ipynb:
```diff
@@ -0,0 +1,419 @@
+                "## Training\n",
+                "\n",
+                "The remainder of this notebook again follows the typical training routine.\n",
+                "With the addition of `CheMeleon` your model may take longer to train but will (hopefully!) have better performance, particularly if the dataset you have is small!"
```
Maybe add some explanation about why the model may take longer to train. I assume it is because the message passing is big, but that you can probably train with fewer epochs because it is already pretrained which makes it quicker.

craabreu replied:
Freezing the message-passing layer may also work well for some targets.

JacksonBurns replied:
clarified both of these points

craabreu commented on chemprop/cli/train.py:
```diff
@@ -497,6 +509,45 @@ def validate_train_args(args):
+                        argument=None,
+                        message=f"Foundation model {fm_name} must be used with `--multi-hot-atom-featurizer-mode V2` not `{mode}`!",
+                    )
+        for arg_value, arg_name in (
+            (args.message_hidden_dim, "--message-hidden-dim"),
+            (args.message_bias, "--message-bias"),
+            (args.depth, "--depth"),
+            (args.undirected, "--undirected"),
+            (args.dropout, "--dropout"),
+            (args.activation, "--activation"),
+            (args.aggregation, "--aggregation"),
+            (args.aggregation_norm, "--aggregation-norm"),
+            (args.atom_messages, "--atom-messages"),
+        ):
+            if arg_value is not None:
+                logger.warning(
+                    f"Passed value `{arg_name} {arg_value}` (possibly default) is overwritten by `--from-foundation {fm_name}`!"
+                )
```
Just a reminder that the `activation` arg is also used to build the FFN.

craabreu commented on chemprop/cli/train.py:
```diff
@@ -918,65 +969,105 @@ def build_model(
+    if from_foundation is not None:
+        match FoundationModels.get(from_foundation):
+            case FoundationModels.CHEMELEON:
+                ckpt_dir = Path(__file__).parent / "_foundation_models"
```
It probably won't break the package installation, but it doesn't look like good practice.

craabreu commented on examples/chemeleon_foundation_finetuning.ipynb:
```diff
@@ -0,0 +1,419 @@
+            "cell_type": "markdown",
+            "metadata": {},
+            "source": [
+                "If you have an existing ChemProp model, you can simply replace your `agg`, `featurizer`, and `mp` with these classes and you can immediately take advantage of `CheMeleon`!"
```
Typo: intialized -> initialized

craabreu commented on examples/chemeleon_foundation_finetuning.ipynb:
```diff
@@ -0,0 +1,419 @@
+                "## Standard ChemProp Preparation\n",
+                "\n",
+                "The below code handles importing needed modules, setting up the data, and initializing the ChemProp model.\n",
+                "It's **mostly** the same as the `training` example provided in the ChemProp repository - for a more detailed breakdown, check that notebook.\n",
```
Adding a link to the mentioned notebook on ReadTheDocs would be nice.

craabreu replied:
It's also worth adding a note about using CheMeleon for classification tasks and a link to the corresponding example notebook.

JacksonBurns replied:
Done in f79e33c2bcaeff635cfe8270c445183fe2c4d9b8 and fe79bf70062f5e80c5c354e1238e41fb1be3f0e5

JacksonBurns commented:
@KnathanM @craabreu @akshatzalte I have addressed all points - please re-review! Of special note:
 - local models for foundation finetuning are now possible by simply specifying a filepath (test added)
 - the notebook has been clarified
 - the model weight saving approach has been changed to allow use of `weights_only=True`

craabreu commented on examples/chemeleon_foundation_finetuning.ipynb:
```diff
@@ -0,0 +1,438 @@
+                "It's **mostly** the same as the `training` example provided in the Chemprop repository - for a more detailed breakdown, [see the docs here](https://chemprop.readthedocs.io/en/latest/training.html).\n",
+                "\n",
+                "The one important change is that we must set `input_dim=mp.output_dim` when we initialize our FFN.\n",
+                "This ensure that the dimension of the learned representation from `CheMeleon` matches the input size for the regressor.\n",
```
```suggestion
                "This ensures that the dimension of the learned representation from `CheMeleon` matches the input size for the regressor.\n",
```

KnathanM commented on chemprop/cli/train.py:
```diff
@@ -918,65 +969,105 @@ def build_model(
+                else:
+                    mp_block = torch.load(
+                        ckpt_dir / "chemeleon_mp.pt", map_location="cpu", weights_only=False
+                    )
```
That is interesting that our default for activation is the Chemporp enumeration. We may change that in a future version to be the string instead. Thank you for including all these details.

KnathanM commented on chemprop/cli/train.py:
```diff
@@ -145,6 +153,10 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
     )
+    transfer_args.add_argument(
+        "--from-foundation",
+        help=f"Name of pretrained foundation model used to initialize message passing. One of: {', '.join((FoundationModels.keys()))}",
```
```suggestion
        help=f"Name of pretrained foundation model used to initialize message passing. One of: {', '.join((FoundationModels.keys()))}, or a path to a local model file.",
```
This could be updated to let users know they can pass their own model.

KnathanM reviewed (changes requested):
Thanks for all your work on this. The PR is looking very good. Just two small changes.

JacksonBurns commented on examples/chemeleon_foundation_finetuning.ipynb:
```diff
@@ -0,0 +1,438 @@
+                "If you have existing Chemprop training code you can simply replace your `agg`, `featurizer`, and `mp` with these classes and you can immediately take advantage of `CheMeleon`!\n",
+                "\n",
+                "In general, we suggest continuing to train the `CheMeleon` weights during finetuning.\n",
+                "You may find that in some cases, freezing the weights (`mp.eval()`) may improve performance."
```
Ah thanks for the catch, I remember you showing me this some time ago! PyTorch quirks abound

JacksonBurns commented:
@KnathanM thanks for the additional fixes! All should be good to go now
